### PR TITLE
[PotentialFlow] Refactoring adjoint element

### PR DESCRIPTION
**Pull request by marcnunezc**
Hi,

This PR aims to make the implementation of #4686 simpler. We are moving from having only one "adjoint_element" that uses finite differences to compute the SensitivitiyMatrix, to a scheme where each non-base element will implement its SensitivityMatrix:
Untitled Diagram

In this PR only the FD element is implemented. The analytical incompressible element is part of #4686.

### CODE REVIEW

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h [hunk 1]
```diff
@@ -90,12 +90,6 @@ class AdjointPotentialFlowElement : public Element
     ///@name Operations
     ///@{
 
-    Element::Pointer Create(IndexType NewId, NodesArrayType const& ThisNodes, PropertiesType::Pointer pProperties) const override;
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h:93]: Here I remove the Create() and Clone() functions of the base element since it is not meant to be used by itself. Is this allowed?

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h [hunk 2]
```diff
@@ -140,30 +134,25 @@ class AdjointPotentialFlowElement : public Element
...
-   Element::Pointer pGetPrimalElement();
-
+    Element::Pointer pGetPrimalElement();
 
 protected:
 
-
     Element::Pointer mpPrimalElement;
 
-    void GetWakeDistances(array_1d<double,NumNodes>& distances);
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h:151]: I remove here GetWakeDistances() because I now use the function from potential_flow_utilities

File: applications/CompressiblePotentialFlowApplication/compressible_potential_flow_application.h [hunk 3]
```diff
@@ -101,8 +101,8 @@ class KRATOS_API(COMPRESSIBLE_POTENTIAL_FLOW_APPLICATION) KratosCompressiblePote
 
     const IncompressiblePotentialFlowElement<2,3> mIncompressiblePotentialFlowElement2D3N;
     const CompressiblePotentialFlowElement<2,3> mCompressiblePotentialFlowElement2D3N;
-    const AdjointPotentialFlowElement<IncompressiblePotentialFlowElement<2,3>> mAdjointIncompressiblePotentialFlowElement2D3N;
-    const AdjointPotentialFlowElement<CompressiblePotentialFlowElement<2,3>> mAdjointCompressiblePotentialFlowElement2D3N;
+    const AdjointFiniteDifferencePotentialFlowElement<IncompressiblePotentialFlowElement<2,3>> mAdjointIncompressiblePotentialFlowElement2D3N;
```
Comment (marcnunezc) [inline applications/CompressiblePotentialFlowApplication/compressible_potential_flow_application.h:104]: When defining the element in the application now we use the FiniteDifference element

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h (hunk 1 above)
Comment (inigolcanalejo) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h:93]: I don't know, let me ping @philbucher and @msandre

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp [hunk 4]
```diff
@@ -13,108 +13,85 @@
...
         mpPrimalElement->Initialize();
     }
 
     template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo) 
+    void AdjointBasePotentialFlowElement<TPrimalElement>::InitializeSolutionStep(ProcessInfo& rCurrentProcessInfo)
     {
         mpPrimalElement->Data() = this->Data();
         mpPrimalElement->Set(Flags(*this));
         mpPrimalElement->InitializeSolutionStep(rCurrentProcessInfo);
```
Comment (philbucher) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:38]: why is this done in `InitializeSolutionStep`?

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:38]: I added this in my first adjoint PR , since the "mpPrimalElement" depends on some flags to be properly solved. Since this "mpPrimalElement" (that comes from the template)  is empty, here I copy flags/data needed. 

I do this in the InitialiazeSolutionStep because the flags/data are set in other processes in the ExecuteInitialize/ExecuteInitializeSolutionStep.

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+template <class TPrimalElement>
+class AdjointFiniteDifferencePotentialFlowElement : public AdjointBasePotentialFlowElement<TPrimalElement>
+{
+public:
+
+    ///@name Type Definitions
+    ///@{
+
+    typedef AdjointBasePotentialFlowElement<TPrimalElement> BaseType;
+    static constexpr int NumNodes = TPrimalElement::TNumNodes;
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h]: Both NumNodes and Dim are already defined in the base class. I think it is better not to redifine them here but instead use:

```suggestion
    using BaseType::NumNodes;
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    static constexpr int Dim = TPrimalElement::TDim;
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h]:
```suggestion
    using BaseType::Dim;
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    KRATOS_CLASS_INTRUSIVE_POINTER_DEFINITION(AdjointFiniteDifferencePotentialFlowElement);
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /**
+     * Constructor.
+     */
+    AdjointFiniteDifferencePotentialFlowElement(IndexType NewId = 0)
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h]: [https://stackoverflow.com/a/121163](https://stackoverflow.com/a/121163)

```suggestion
    explicit AdjointFiniteDifferencePotentialFlowElement(IndexType NewId = 0): AdjointBasePotentialFlowElement<TPrimalElement>(NewId){};
```

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.cpp [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+    }
+
+
+    template <class TPrimalElement>
+    void AdjointFiniteDifferencePotentialFlowElement<TPrimalElement>::CalculateSensitivityMatrix(const Variable<array_1d<double,3> >& rDesignVariable,
+                                            Matrix& rOutput,
+                                            const ProcessInfo& rCurrentProcessInfo)
+    {
+        KRATOS_TRY;
+        const double delta = this->GetPerturbationSize();
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.cpp:51]: I think we discussed this in another PR, but I think that we are mixing the finite difference from the element and the response function.  Would it be better to pass the PerturbationSize within the rCurrentProcessInfo? And then do something like:

```suggestion
        const double delta = rCurrentProcessInfo[SCALE_FACTOR];
```

Also, maybe a better naming for the variable might be a good idea. In the structural mechanics application they do:

```cpp
const double perturbation_size = rCurrentProcessInfo[PERTURBATION_SIZE];
```

This would be anyway for another PR.

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h [hunk 9]
```diff
@@ -140,30 +134,25 @@ class AdjointPotentialFlowElement : public Element
...
-
     void GetValuesOnSplitElement(Vector& split_element_values, const array_1d<double,NumNodes>& distances);
 
+    double GetPerturbationSize();
```
Comment (inigolcanalejo) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h]: I would actually implement this function as a private function only in the AdjointFiniteDifferencePotentialFlowElement since it is only used there.

**Review by inigolcanalejo — CHANGES_REQUESTED** [review]

**Comment by inigolcanalejo** [comment]
Hi @marcnunezc, great to see the new structure of the adjoint elements! :+1: I only have the minor above comments. Short question: shall we merge this before or after https://github.com/KratosMultiphysics/Kratos/pull/4686?

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h (hunk 9 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h]: Absolutely!

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.cpp (hunk 8 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.cpp:51]: You are right! I would leave this for another PR since I also wanted to add the possibility to define the perturbation size in terms of the nodal size

**Comment by marcnunezc** [comment]
> Hi @marcnunezc, great to see the new structure of the adjoint elements! +1 I only have the minor above comments. Short question: shall we merge this before or after #4686?

Thanks for your comments! My idea was to merge this first so that you can adapt your PR later (as long as this PR doesn't take too long). Is that ok for you?

**Comment by inigolcanalejo** [comment]
> > Hi @marcnunezc, great to see the new structure of the adjoint elements! +1 I only have the minor above comments. Short question: shall we merge this before or after #4686?
> 
> Thanks for your comments! My idea was to merge this first so that you can adapt your PR later (as long as this PR doesn't take too long). Is that ok for you?

Sounds good!

**Review by inigolcanalejo — DISMISSED** [review]

**Comment by inigolcanalejo** [comment]
I removed my blocker, from my point of view it is ready for merging.

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp [hunk 10]
```diff
@@ -13,108 +13,85 @@
...
     }
 
     template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
+    void AdjointBasePotentialFlowElement<TPrimalElement>::CalculateLocalSystem(MatrixType& rLeftHandSideMatrix,
                                       VectorType& rRightHandSideVector,
-                                      ProcessInfo& rCurrentProcessInfo) 
+                                      ProcessInfo& rCurrentProcessInfo)
     {
         mpPrimalElement->CalculateLocalSystem(rLeftHandSideMatrix,
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: I guess that `mpPrimalElement` is a pointer to the base element implementing the formulation (incompressible / compressible potential flow elements). If so it is, do you need a pointer to it? Isn't it sufficient to do `TPrimalElement::CalculateLocalSystem()`?

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp [hunk 11]
```diff
@@ -13,108 +13,85 @@
...
         mpPrimalElement->GetValueOnIntegrationPoints(rVariable,rValues,rCurrentProcessInfo);
     }
 
     template <class TPrimalElement>
-    void AdjointPotentialFlowElement<TPrimalElement>::GetValuesVector(Vector& rValues, int Step) 
+    void AdjointBasePotentialFlowElement<TPrimalElement>::GetValuesVector(Vector& rValues, int Step)
     {
         KRATOS_TRY
-        const AdjointPotentialFlowElement& r_this = *this;
+        const AdjointBasePotentialFlowElement& r_this = *this;
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: Why do you need to make it `const`? Is it to have a threadsafe access?

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h (hunk 1 above)
Comment (rubenzorrilla) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h:93]: If the base element can't be standalone used I think it makes sense to remove them.

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.cpp [hunk 12]
```diff
@@ -0,0 +1,134 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.cpp]: Header format

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h [hunk 13]
```diff
@@ -0,0 +1,120 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
```
Comment (rubenzorrilla) [inline applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_finite_difference_potential_flow_element.h]: Header format

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 11 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: Yes! This was implemented by @inigolcanalejo  in the incompressible element follwoing suggestions by @philbucher (if im not mistaken). @philbucher do you have more insight on this?

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 10 above)
Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: So this follows @armingeiser and @msandre design if im not mistaken, please correct me if im wrong:

TPrimalElement is not constructed (with an index, a geometry, etc) so we can not call directly the methods from it, as it has no data.

Here is where the adjoint element is defined:

https://github.com/KratosMultiphysics/Kratos/blob/1df0f52127ddbaf9cdbbef62c368fe4bf9540fcd/applications/CompressiblePotentialFlowApplication/compressible_potential_flow_application.h#L104-L105

At the constructor of the adjoint element, we construct the primal element and  we store a pointer (once initiazlied with an index, geometry, etc..):

https://github.com/KratosMultiphysics/Kratos/blob/1df0f52127ddbaf9cdbbef62c368fe4bf9540fcd/applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.h#L53-L58

So to answer your question, I would say that it is not possible to use `TPrimalElement::CalculateLocalSystem()` since TPrimalElement hasn't been constructed yet.

Comment (armingeiser) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: Right, `TPrimalElement` is just the class, but the function needs to be called at the instance -  `TPrimalElement` is **not** the base element.

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 11 above)
Comment (rubenzorrilla) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: The point is that you are accessing to the element database. Provided that you've initialized the variables you are accessing to, there shouldn't be any problem. @philbucher am I wrong?

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 10 above)
Comment (rubenzorrilla) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: Thanks for the explanation.

Now I see. I though you followed a similar design to the ones in the FluidDynamics, in which we derive from a base element that implements the formulation instead of deriving from the base `Element` in the core.

Just for me to know, is there any reason to not derive from the compressible / incompressible elements? Even though the elements are not shared pointers anymore, it may be still worth for the sake of memory consumption.

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: The main reason is that I followed the StructuralApp design to implement the adjoint in the potential flow, especially to use their StructuralAdjoint utilitities (sensitivity builder) and processes (replace multiple element and conditions). Now we are using both processes from the core so a different scheme could be considered. Also the ShapeOpt examples were running with the structural case so it was more straight forward to follow the structural approach.

Deriving from the compressible/incomressible elements (and the embedded ones) would require to have a complete adjoint for each of those if Im not mistaken (with repeated code). That would be (in the short term) 4 primal elements + 4 adjoint elements, which is not too much actually. It wouldn't be too much work either since the adjoint elements do not have many functions.

The advantage of the current approach is that a new element formulation would work out of the box with the same adjoint element without having to implement anything else. For instance, once we added the compressible element, we didn't have to add new functions to the adjoint element since it could use the same one that was currently implemented. We only needed to define it in the application with its correct template.

In short I think that the current implementation is better and cleaner if more elements/formulations are planned to be added. Deriving from the primals direclty could be considered however to improve performance if it is worth to do so.

Not sure if there is another big reason to use template `TPrimalElement` instead of deriving from the primals @armingeiser ?

Comment (rubenzorrilla) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: What I mean is that by doing
```c++
template <class TPrimalElement>
class AdjointElement : public TPrimalElement {...}
```
instead of 
```c++
template <class TPrimalElement>
class AdjointElement : public Element {...}
```
you could save the overhead of one pointer per element with no behavior change. You will only need to forward declare for each one of the primal elements, as you are doing right now.

I was just wondering. There should be a reason because of which you selected the current design.

Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: I haven't followed the discussion closely but I remember discussion this many times with @armingeiser 
afair the first version of the adjoint elements did derive from the primal version, but there were some issues that could not really be solved (@armingeiser maybe you remember what it was?)
Hence the decision was taken to change to the current approach, which to my knowledge works best so far
also ping @msandre since he was also involved in this design

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 11 above)
Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: @rubenzorrilla you are absolutely right
=> accessing a variable of the element will [allocate this variable if it does not exist yet](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/containers/data_value_container.h#L223) (=> this is the correct behavior of the `DataValueContainer`), which might not always be desired. This ofc cannot happen with a const version of the element

I had in mind to clearly tell others that take a look at this code: Here I do NOT want to allocate anything.

Ofc it also works if I correctly allocate the memory beforehand, but this is done outside of the elements. Now someone writes a new solver, then the behavior changes unknowingly.

(BTW not so long ago I had to debug sth for quite some time exactly bcs of what I described :) )

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 10 above)
Comment (armingeiser) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: Yes. In our case we can not derive from the primal element, because we need the primal to work as is - It's a bit complicated to explain, but i try:
The main reason is that the adjoint sensitivity analysis is composed of the adjoint problem (adjoint elements) itself and the derivatives of the system (primal elements). In the current design all this is done using the same model part, that only has the adjoint elements. This gets complicated by the fact that we are calculating most of the derivatives with finite differences, e.g. we need to be able to calculate stresses in the primal element using the primal variables as dofs, while the adjoint elements have the adjoint variables as dofs. If an adjoint element derives from the primal element and overwrites the GetDofs, GetValues methods, in the example given, the stresses would be calculated using the `ADJOINT_DISPLACEMENT`.

Comment (rubenzorrilla) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp]: Thanks for the clear explanation @armingeiser . I think that more or less I got the idea. Honestly, no better design comes to my mind neither.

File: applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp (hunk 11 above)
Comment (rubenzorrilla) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: Long story short, in here is that you don't want to allocate the memory and initialize the value in the `Initialize()` of the element but to assume that it has been done beforehand by any other utility. Am I wrong?

May I ask why you don't want this allocation to be done? (just for me to know).

Comment (marcnunezc) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: I encountered some execution errors in the current master when running in FullDebug when using non historical variables. I have "fixed" them by using the same technique as here, but im still not clear if this is the intended way to do so. 

Since this discussion goes a bit further than the scope of this PR I suggest following the discussion in #5189 , where I  pushed the hotfix.

Comment (philbucher) [reply applications/CompressiblePotentialFlowApplication/custom_elements/adjoint_base_potential_flow_element.cpp:94]: > May I ask why you don't want this allocation to be done? (just for me to know).

I am not saying that I do not want the allocation, but I want to point out the behavior since it can affect how you write the code.
It really depends on the usecase

**Comment by inigolcanalejo** [comment]
Is this ready for merging?

**Comment by marcnunezc** [comment]
Let me resolve conflicts and check that all tests pass after #5189!

**Comment by marcnunezc** [comment]
This one should be ready to be merged!

**Review by inigolcanalejo — APPROVED** [review]
